bindings/wasm/src/resolver/wasm_resolver.rs

```diff
@@ -237,26 +230,53 @@ impl WasmResolver {
     presentation: &WasmPresentation,
     options: &WasmPresentationValidationOptions,
     fail_fast: WasmFailFast,
-    holder: &OptionSupportedDocument,
-    issuers: &OptionArraySupportedDocument,
+    holder: &OptionIAsCoreDocument,
+    issuers: &OptionArrayIAsCoreDocument,
   ) -> Result<PromiseVoid> {
-    let resolver: Rc<RustSupportedDocumentResolver> = self.0.clone();
+    let resolver: Rc<JsDocumentResolver> = self.0.clone();
     let presentation: Presentation = presentation.0.clone();
     let options: PresentationValidationOptions = options.0.clone();
 
-    let holder: Option<RustSupportedDocument> = holder.into_serde().wasm_result()?;
-    let holder: Option<RustSupportedDocument> = holder.map(From::from);
-    let issuers: Option<Vec<RustSupportedDocument>> = issuers.into_serde().wasm_result()?;
+    let holder_clone: JsValue = JsValue::from(holder);
+    let issuers_clone: JsValue = JsValue::from(issuers);
+
     let promise: Promise = future_to_promise(async move {
-      resolver
-        .verify_presentation(
-          &presentation,
-          &options,
-          fail_fast.into(),
-          holder.as_ref(),
-          issuers.as_deref(),
-        )
-        .await
+      let holder_handler = || async {
+        if holder_clone.is_null() || holder_clone.is_undefined() {
+          resolver
+            .resolve_presentation_holder(&presentation)
+            .await
+            .map(|value| ImportedDocumentLock::from_js_value_unchecked(&value))
+        } else {
+          Ok(ImportedDocumentLock::from_js_value_unchecked(&holder_clone))
+        }
+      };
+
+      let issuers_handler = || async {
+        if issuers_clone.is_null() || issuers_clone.is_undefined() {
+          resolver.resolve_presentation_issuers(&presentation).await.map(|value| {
+            value
+              .into_iter()
+              .map(|entry| ImportedDocumentLock::from_js_value_unchecked(&entry))
+              .collect::<Vec<ImportedDocumentLock>>()
+          })
+        } else {
+          Ok(Vec::<ImportedDocumentLock>::from(
+            issuers_clone.unchecked_ref::<ArrayIAsCoreDocument>(),
+          ))
+        }
+      };
+
+      let (holder_result, issuers_result) = futures::join!(holder_handler(), issuers_handler());
+
+      let holder = holder_result.wasm_result()?;
+      let issuers = issuers_result.wasm_result()?;
+
+      let holder_guard = holder.blocking_read();
+      let issuers_guard: Vec<ImportedDocumentReadGuard<'_>> =
+        issuers.iter().map(ImportedDocumentLock::blocking_read).collect();
```

Comment:
Perhaps these should be changed to the async read versions since we are in an async context. Only worth it if we decide not to port the resolver to TS though.

Reply:
> Only worth it if we decide not to port the resolver to TS though.

What do you mean here? We are already porting the resolver to TS, aren't we?

Reply:
I think it would be valuable to write a resolver in TS instead of using `WasmResolver`.  Note that almost all the methods on `WasmResolver` have quite a bit of code and it isn't as simple as just forwarding to the Rust implementation as is the case with most other types in the bindings. Since there are several other advantages with that approach on both the Rust and TS side I would recommend it.

Reply:
Aside/Regardless: I've changed the locking to use the async versions here now.